ndk/src/event.rs

```diff
@@ -373,7 +398,26 @@ impl MotionEvent {
         Self { ptr }
     }
 
-    /// Returns a pointer to the native [`ffi::AInputEvent`]
+    /// Creates a native [`InputEvent`] object that is a copy of the specified
+    /// Java [`android.view.MotionEvent`]. The result may be used with generic and
+    /// [`MotionEvent`]-specific functions.
+    ///
+    /// # Safety
+    ///
+    /// This function should be called with a healthy JVM pointer and with a non-null
+    /// [`android.view.MotionEvent`], which must be kept alive on the Java/Kotlin side.
```

Comment:
@rib I copied the `which must be kept alive on the Java/Kotlin side.` portion from other `from_java()` functions, but I don't think it holds true for these APIs.

1. As the copied doc-comment above suggests: `... that is a copy of the specified Java ...`;
2. And `AInputEvent_release()` which must be called states: `The underlying Java object remains valid and does not change its state.`.

It seems the two objects are independent and have their own lifetime after this call.

If you agree I'll remove this text.

Reply:
Hmm, I'm not sure tbh - from when I last tried to get my head around the implementation details for input event handling I recall some pretty funky lifetime tracking for events that went via an `InputQueue` since there's some special callbacks in java for events after they are handled by native code from what I vaguely recall and I don't really know how the bits all fit together if using these without really digging into the implementation.

My guess would be that you do actually need to ensure the event is kept alive in the jvm if they only hold a weak reference in the native event and guess that the release isn't just about freeing the struct but also tracking when the event is finished with on the jvm side so that it can progress to the next stage of being handled in the jvm.

I'd er on the side of keeping that requirement in the docs unless we really look into the implementation details more closely.

Reply:
It constructs it here:

https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/native/android/input.cpp;l=90;drc=f5c06a171c6eb51c6c914540b36a8211036fe735

Which copies over a bunch of fields:

https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/jni/android_view_KeyEvent.cpp;l=116-139;drc=ea4a158fa1e4620658dab38305ddde79700843f7

All information is copied as raw `jint`s by value, and I don't think the various `id` fields have a back-reference to the Java representation.

---

Likewise the motion events are fully copied:

https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/native/android/input.cpp;l=312-314;drc=f5c06a171c6eb51c6c914540b36a8211036fe735

https://cs.android.com/android/platform/superproject/main/+/main:frameworks/native/libs/input/Input.cpp;l=547-580;drc=ea4a158fa1e4620658dab38305ddde79700843f7

With nothing that appears to be a reference.

---

Let's remove the doc text.

Reply:
> and guess that the release isn't just about freeing the struct but also tracking when the event is finished with on the jvm side so that it can progress to the next stage of being handled in the jvm.

To confirm, the terribly named `_release` (why not `_destroy`... release sounds like refcounting) function only calls `delete` and doesn't perform any "event finished" handling later. That, after all, happens elsewhere via `AInputQueue_finishEvent()` which has a reference to the input queue and exists specifically for explicitness of this purpose.

https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/native/android/input.cpp;l=57-59;drc=f5c06a171c6eb51c6c914540b36a8211036fe735

---

Looks like I already removed the wording in the last force-push, let's merge this.